chain/client/src/client.rs

```diff
@@ -2463,42 +2562,97 @@ impl Client {
         let mut notify_state_sync = false;
         let me = signer.as_ref().map(|x| x.validator_id().clone());
 
-        for (sync_hash, state_sync_info) in self.chain.chain_store().iterate_state_sync_infos()? {
-            assert_eq!(sync_hash, state_sync_info.epoch_tail_hash);
-            let network_adapter = self.network_adapter.clone();
+        for (epoch_first_block, mut state_sync_info) in
```

Comment:
this method is too large - can you split it down into smaller methods please?

Reply:
extracted a get_catchup_sync_hash() function in f96f76203a4c802466f4c14bfb8bdabc5ca4635b